File: swifty_crypto/session.py
"""Confidential session channel — WMK delivery into the CVM (identity-envelope.md §6).

The working-memory key (WMK) is the one user-held key that ever enters a remote machine. It
does so ONLY over a session key `SK` that is bound to the verified attestation:

  1. Client generates an ephemeral X25519 keypair; `client_eph_pub` is the attestation
     challenge, so the CVM's quote binds `SHA-256(client_eph_pub)` in report_data
     (attestation.md §3-4) — the keystone verifies this BEFORE any key leaves the client.
  2. Both sides compute `SK = HKDF(X25519(·), info = "swifty/session/v1" ‖ client_eph_pub ‖
     cvm_pub)`. Binding BOTH public keys into the KDF info defeats unknown-key-share / a MITM
     who substitutes a key: a substituted key yields a different `SK`, so the sealed WMK
     won't open.
  3. Client seals WMK to `SK` (XChaCha20 envelope, purpose "wmk-delivery") and sends it; the
     CVM opens it in TEE RAM, uses it for the session, and discards it on session end (§6).

X25519 here matches CryptoKit's `Curve25519.KeyAgreement` and the gateway's `cryptography`
X25519 (all RFC 7748), so client and CVM derive the SAME SK.
"""

from __future__ import annotations

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from swifty_crypto.envelope import open as _open
from swifty_crypto.envelope import seal as _seal

SESSION_LABEL = b"swifty/session/v1"
WMK_DELIVERY_PURPOSE = "wmk-delivery"
KEY_LEN = 32


def x25519_keypair() -> tuple[bytes, bytes]:
    """Return (private_bytes[32], public_bytes[32]) for an ephemeral session key."""
    priv = X25519PrivateKey.generate()
    return (
        priv.private_bytes_raw(),
        priv.public_key().public_bytes_raw(),
    )


def x25519_public(private_bytes: bytes) -> bytes:
    return X25519PrivateKey.from_private_bytes(private_bytes).public_key().public_bytes_raw()


def derive_session_key(my_private: bytes, their_public: bytes, *, client_pub: bytes, cvm_pub: bytes) -> bytes:
    """SK from ECDH + HKDF, binding both endpoints' public keys (anti-UKS/MITM).

    Both sides pass the SAME `client_pub`/`cvm_pub` (the channel identities) and their own
    `my_private` + the peer's `their_public`; the ECDH is symmetric so they agree on SK."""
    if len(client_pub) != KEY_LEN or len(cvm_pub) != KEY_LEN:
        raise ValueError("public keys must be 32 bytes")
    if len(my_private) != KEY_LEN or len(their_public) != KEY_LEN:
        raise ValueError("session keys must be 32 bytes")
    # `cryptography`'s exchange() raises on an all-zero (low-order/identity) result, so a
    # substituted low-order peer key fails closed here (the Swift side mirrors this).
    shared = X25519PrivateKey.from_private_bytes(my_private).exchange(
        X25519PublicKey.from_public_bytes(their_public)
    )
    info = SESSION_LABEL + client_pub + cvm_pub  # fixed-length pubs -> unambiguous
    return HKDF(algorithm=hashes.SHA256(), length=KEY_LEN, salt=None, info=info).derive(shared)


def seal_wmk(session_key: bytes, wmk: bytes, *, account_id: str, v: int = 1) -> dict:
    """Client seals WMK to the session key for delivery into the CVM (§6)."""
    return _seal(session_key, wmk, account_id=account_id, purpose=WMK_DELIVERY_PURPOSE, v=v)


def open_wmk(session_key: bytes, envelope: dict, *, account_id: str) -> bytes:
    """CVM opens the sealed WMK (in TEE RAM). Fails closed on any tamper / wrong session."""
    return _open(session_key, envelope, account_id=account_id, purpose=WMK_DELIVERY_PURPOSE)

File: caladon-core/src/session.rs
//! Confidential session channel — WMK delivery into the CVM (identity-envelope.md §6).
//! Byte-identical to `swifty_crypto/session.py` and `SwiftyKit/Session.swift`; all three derive
//! the SAME session key so client and CVM agree (X25519 / RFC 7748 + HKDF-SHA256).
//!
//! The working-memory key (WMK) is the one user-held key that ever enters a remote machine, and
//! only over a session key `SK` bound to the verified attestation. `SK = HKDF(X25519(·),
//! info = "swifty/session/v1" ‖ client_pub ‖ cvm_pub)` — binding BOTH endpoints' public keys
//! into the KDF info defeats unknown-key-share / a MITM key substitution.
//!
//! CRITICAL: x25519-dalek does NOT raise on a low-order/identity peer key (the resulting shared
//! secret is all-zero); Python's `cryptography` DOES (`exchange()` raises), and Swift mirrors it.
//! We must reject explicitly here (fail-closed) or a substituted low-order key yields an
//! attacker-determined SK with no ECDH entropy.

use hkdf::Hkdf;
use sha2::Sha256;
use x25519_dalek::{PublicKey, StaticSecret};

use crate::envelope::{self, EnvelopeError};

pub const SESSION_LABEL: &[u8] = b"swifty/session/v1";
pub const WMK_DELIVERY_PURPOSE: &str = "wmk-delivery";
pub const CHAT_PURPOSE: &str = "chat";
pub const KEY_LEN: usize = 32;

#[derive(Debug, PartialEq, Eq)]
pub enum SessionError {
    BadPublicKeyLength,
    BadPrivateKeyLength,
    /// Peer key was low-order/identity (all-zero shared secret) — fail closed.
    LowOrderPoint,
    Envelope(EnvelopeError),
}

impl From<EnvelopeError> for SessionError {
    fn from(e: EnvelopeError) -> Self {
        SessionError::Envelope(e)
    }
}

/// X25519 public key (32 bytes) for an ephemeral/long-term private scalar. The scalar is clamped
/// during the base-point multiply (RFC 7748), matching CryptoKit + `cryptography`.
pub fn x25519_public(private_bytes: &[u8]) -> Result<[u8; 32], SessionError> {
    let priv_arr: [u8; 32] = private_bytes
        .try_into()
        .map_err(|_| SessionError::BadPrivateKeyLength)?;
    let secret = StaticSecret::from(priv_arr);
    Ok(PublicKey::from(&secret).to_bytes())
}

/// SK = HKDF(X25519(my_private, their_public), info = label ‖ client_pub ‖ cvm_pub).
/// Both sides pass the SAME client_pub/cvm_pub (the channel identities) and their own private +
/// the peer's public; the ECDH is symmetric so they agree on SK.
///
/// HKDF here is salt=None (RFC-5869 all-zero salt), matching Python's `HKDF(salt=None)` and
/// Swift's `hkdfDerivedSymmetricKey(salt: Data())`.
pub fn derive_session_key(
    my_private: &[u8],
    their_public: &[u8],
    client_pub: &[u8],
    cvm_pub: &[u8],
) -> Result<[u8; KEY_LEN], SessionError> {
    if client_pub.len() != KEY_LEN || cvm_pub.len() != KEY_LEN {
        return Err(SessionError::BadPublicKeyLength);
    }
    let priv_arr: [u8; 32] = my_private
        .try_into()
        .map_err(|_| SessionError::BadPrivateKeyLength)?;
    let pub_arr: [u8; 32] = their_public
        .try_into()
        .map_err(|_| SessionError::BadPublicKeyLength)?;

    let secret = StaticSecret::from(priv_arr);
    let peer = PublicKey::from(pub_arr);
    let shared = secret.diffie_hellman(&peer);

    // Reject a low-order/identity peer point. `was_contributory()` is false exactly when the
    // shared secret is the all-zero point; we also check the bytes directly as belt-and-braces.
    if !shared.was_contributory() || shared.as_bytes() == &[0u8; 32] {
        return Err(SessionError::LowOrderPoint);
    }

    let info: Vec<u8> = [SESSION_LABEL, client_pub, cvm_pub].concat();
    let hk = Hkdf::<Sha256>::new(None, shared.as_bytes());
    let mut sk = [0u8; KEY_LEN];
    hk.expand(&info, &mut sk)
        .expect("hkdf expand within output bound");
    Ok(sk)
}

/// Client seals WMK to the session key for delivery into the CVM (§6). Deterministic-nonce
/// parity entry point; production should use `seal_wmk` (random nonce).
pub fn seal_wmk_with_nonce(
    session_key: &[u8],
    wmk: &[u8],
    account_id: &str,
    v: i64,
    nonce: &[u8],
) -> Result<Vec<u8>, SessionError> {
    Ok(envelope::seal_with_nonce(
        session_key,
        wmk,
        account_id,
        WMK_DELIVERY_PURPOSE,
        v,
        nonce,
    )?)
}

/// Client seals WMK to the session key (random nonce, production path). Returns `(nonce, ct)`.
pub fn seal_wmk(
    session_key: &[u8],
    wmk: &[u8],
    account_id: &str,
    v: i64,
) -> Result<([u8; envelope::NONCE_LEN], Vec<u8>), SessionError> {
    Ok(envelope::seal(
        session_key,
        wmk,
        account_id,
        WMK_DELIVERY_PURPOSE,
        v,
    )?)
}

/// CVM opens the sealed WMK (in TEE RAM). Fails closed on any tamper / wrong session.
pub fn open_wmk(
    session_key: &[u8],
    nonce: &[u8],
    ct: &[u8],
    account_id: &str,
    v: i64,
) -> Result<Vec<u8>, SessionError> {
    Ok(envelope::open_bound(
        session_key,
        nonce,
        ct,
        account_id,
        WMK_DELIVERY_PURPOSE,
        v,
    )?)
}

/// Seal a live-turn payload (prompt or response delta) to SK (purpose "chat").
pub fn seal_chat_with_nonce(
    session_key: &[u8],
    plaintext: &[u8],
    account_id: &str,
    v: i64,
    nonce: &[u8],
) -> Result<Vec<u8>, SessionError> {
    Ok(envelope::seal_with_nonce(
        session_key,
        plaintext,
        account_id,
        CHAT_PURPOSE,
        v,
        nonce,
    )?)
}

/// Random-nonce chat seal (production path). Returns `(nonce, ct)`.
pub fn seal_chat(
    session_key: &[u8],
    plaintext: &[u8],
    account_id: &str,
    v: i64,
) -> Result<([u8; envelope::NONCE_LEN], Vec<u8>), SessionError> {
    Ok(envelope::seal(session_key, plaintext, account_id, CHAT_PURPOSE, v)?)
}

/// Open a sealed live-turn payload under SK. Fails closed on tamper.
pub fn open_chat(
    session_key: &[u8],
    nonce: &[u8],
    ct: &[u8],
    account_id: &str,
    v: i64,
) -> Result<Vec<u8>, SessionError> {
    Ok(envelope::open_bound(
        session_key,
        nonce,
        ct,
        account_id,
        CHAT_PURPOSE,
        v,
    )?)
}
